adapter/pkg/synchronizer/client_pool.go

```diff
@@ -0,0 +1,188 @@
+	for workerReq := range w.internalQueue {
+		responseReceived := w.processFunc(&workerReq.Req, workerReq.APIUUID, workerReq.labels, workerReq.SyncAPIRespChannel,
+			&workerPool.client)
+		if !responseReceived {
+			time.Sleep(w.delayAfterFault)
+		}
+	}
+}
+
+var (
+	// WorkerPool is the thread pool responsible for sending the control plane request to fetch APIs
+	workerPool        *pool
+	oncePoolInitiated sync.Once
+)
+
+// InitializeWorkerPool creates the Worker Pool used for the Control Plane Rest API invocations.
+// maxWorkers indicate the maximum number of parallel workers sending requests to the control plane.
+// jobQueueCapacity indicate the maximum number of requests can kept inside a single worker's queue.
+// delayForFaultRequests indicate the delay a worker enforce (in seconds) when a fault response is received.
+func InitializeWorkerPool(maxWorkers, jobQueueCapacity int, delayForFaultRequests time.Duration, trustStoreLocation string,
+	skipSSL bool, requestTimeout, retryInterval time.Duration, serviceURL, username, password string) {
+	// TODO: (VirajSalaka) Think on whether this could be moved to global adapter seamlessly.
+	oncePoolInitiated.Do(func() {
+		workerPool = newWorkerPool(maxWorkers, jobQueueCapacity, delayForFaultRequests)
+		workerPool.controlPlaneParams = controlPlaneParameters{
+			serviceURL:    serviceURL,
+			username:      username,
+			password:      password,
+			retryInterval: retryInterval,
+		}
+		var tr *http.Transport
+		if !skipSSL {
+			caCertPool := tlsutils.GetTrustedCertPool(trustStoreLocation)
+			tr = &http.Transport{
+				TLSClientConfig: &tls.Config{RootCAs: caCertPool},
+			}
+		} else {
+			tr = &http.Transport{
+				TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+			}
+		}
+		// Configure Connection Level Parameters since it is reused over and over
+		tr.MaxConnsPerHost = maxWorkers * 2
+		tr.MaxIdleConns = maxWorkers * 2
+		tr.MaxIdleConnsPerHost = maxWorkers * 2
+		workerPool.client = http.Client{
+			Transport: tr,
+			Timeout:   requestTimeout * time.Second,
+		}
+	})
+}
+
+func newWorkerPool(maxWorkers, jobQueueCapacity int, delayForFaultRequests time.Duration) *pool {
+	if jobQueueCapacity <= 0 {
+		jobQueueCapacity = 100
+	}
+	requestChannel := make(chan workerRequest, jobQueueCapacity)
+	workers := make([]*worker, maxWorkers)
+
+	// create workers
+	for i := 0; i < maxWorkers; i++ {
+		workers[i] = &worker{
+			id:              i,
+			internalQueue:   requestChannel,
+			processFunc:     SendRequestToControlPlane,
+			delayAfterFault: delayForFaultRequests,
+		}
+		go workers[i].ProcessFunction()
+		loggers.LoggerSync.Infof("ControlPlane processing worker %d spawned.", i)
+	}
+
+	return &pool{
+		internalQueue: requestChannel,
+		workers:       workers,
+		quit:          make(chan bool),
+	}
+}
+
+// Enqueue Tries to enqueue but fails if queue is full
+func (q *pool) Enqueue(req workerRequest) bool {
+	if len(q.internalQueue) > cap(q.internalQueue)/10*8 {
```

Comment:
Put a comment why this does not hit when a burst of api deploy events are triggered.